lib/spack/spack/environment/environment.py

```diff
@@ -578,7 +603,72 @@ def specs_for_view(self, concretized_root_specs):
 
         return specs
 
-    def regenerate(self, concretized_root_specs):
+    def update_method_error_msg(self, methods):
+        """When methods are already determined invalid, construct error message for methods."""
+        msg = "View cannot be updated using specified update methods:"
+        if "exchange" in methods:
+            if not spack.util.atomic_update.renameat2():
+                msg += "\n  Operating system does not support 'exchange' atomic update method."
+                msg += f"\n    If the view {self.root} does not already exist on the filesystem,"
+                msg += " change its update_method to 'symlink' or 'auto'."
+                msg += f"\n    If the view at {self.root} exists already, either change the"
+                msg += " update_method and run `spack env view regenerate --force`"
+                msg += " or run on a newer OS."
+            else:
+                msg += f"\n  The view {self.root} cannot be updated with 'exchange' update method"
+                msg += " because it was originally constructed with the 'symlink' method."
+                msg += "\n    Either change the update method to 'symlink' or"
+                msg += " run `spack env view regenerate --force` for a non-atomic update."
+        if "symlink" in methods:
+            msg += f"\n  The view {self.root} cannot be updated with 'symlink' update method"
+            msg += " because it was originally constructed with the 'exchange' method."
+            msg += "\n    Either change the update method to 'exchange' or"
+            msg += " run `spack env view regenerate --force` for a non-atomic update."
+        return msg
+
+    def valid_update_method(self, method, force):
+        return getattr(self, f"valid_update_method_{method}")(force)
+
+    def valid_update_method_exchange(self, force):
+        if not spack.util.atomic_update.renameat2():
+            return False
+
+        # This is all to ensure we don't swap symlink -> exchange if we have
+        # symlink as an acceptable method. This is to avoid problems switching
+        # between OSs
+        if os.path.exists(self.root):
+            if os.path.islink(self.root):
+                if force:
+                    os.unlink(self.root)
+                    return True
+                else:
+                    return "symlink" not in self.update_method
+
+        return True
+
+    def valid_update_method_symlink(self, force):
+        if os.path.exists(self.root):
+            if not os.path.islink(self.root):
+                if force:
+                    shutil.rmtree(self.root)
+                    return True
+                else:
+                    return False
+        return True
```

Comment:
Can you improve readability a bit? It's not necessary to have 3 levels of control flow, it can be 1. Also can you add some comments what's being tested for posterity?

```suggestion
        # comment 1
        if not os.path.exists(self.root) or os.path.islink(self.root):
            return True

        # comment 2
        if force:
            shutil.rmtree(self.root)
            return True

       # comment 3
        return False
```

Reply:
Same applies elsewhere.

Reply:
Additionally, the initial check is a complicated way of expresing `lexist`, can you fix that?

Reply:
No, it is not the same as `lexists`. For example, this condition is `False` if the path exists and is a directory, `lexists` returns `True` in that case

Reply:
I think this:

```
    def valid_update_method_exchange(self, force):                                                                                                                                                                              
        if not spack.util.atomic_update.renameat2():                                                                                                                                                                            
            return False                                                                                                                                                                                                        
                                                                                                                                                                                                                                
        if not os.path.islink(self.root):                                                                                                                                                                                       
            return True                                                                                                                                                                                                         
                                                                                                                                                                                                                                
        if force:                                                                                                                                                                                                               
            os.unlink(self.root)                                                                                                                                                                                                
            return True                                                                                                                                                                                                         
        else:                                                                                                                                                                                                                   
            # Ensure we don't swap symlink -> exchange if we have symlink as an acceptable method.                                                                                                                              
            # This is to avoid problems switching between OSs.                                                                                                                                                                  
            return "symlink" not in self.update_method                                                                                                                                                                          
                                                                                                                                                                                                                                
        return True
```

is a lot less legible than

```
    def valid_update_method_exchange(self, force):                                                                                                                                                                              
        if not spack.util.atomic_update.renameat2():                                                                                                                                                                            
            return False                                                                                                                                                                                                        
                                                                                                                                                                                                                                
        # Ensure we don't swap symlink -> exchange if we have a symlink and symlink is an                                                                                                                                       
        # acceptable method. This is to avoid problems switching between OSs.                                                                                                                                                   
        if os.path.islink(self.root):                                                                                                                                                                                           
            if force:                                                                                                                                                                                                           
                os.unlink(self.root)                                                                                                                                                                                            
                return True                                                                                                                                                                                                     
            else:                                                                                                                                                                                                               
                return "symlink" not in self.update_method                                                                                                                                                                      
                                                                                                                                                                                                                                
        return True
```

despite the latter being a level further indented.

Removed the other condition as it was vestigial.

Reply:
And this seems even clearer:

```
    def valid_update_method_exchange(self, force):                                                                                                                                                                              
        if not spack.util.atomic_update.renameat2():                                                                                                                                                                            
            return False                                                                                                                                                                                                        
                                                                                                                                                                                                                                
        # Ensure we don't swap symlink -> exchange if we have a symlink and symlink is an                                                                                                                                       
        # acceptable method. This is to avoid problems switching between OSs.                                                                                                                                                   
        if os.path.islink(self.root):                                                                                                                                                                                           
            if force:                                                                                                                                                                                                           
                os.unlink(self.root)                                                                                                                                                                                            
            elif "symlink" in self.update_method:                                                                                                                                                                               
                return False                                                                                                                                                                                                    
                                                                                                                                                                                                                                
        return True
```

Reply:
I'd have gone with this one:

```python
    def valid_update_method_exchange(self, force):
        if not spack.util.atomic_update.renameat2():
            return False

        if not os.path.islink(self.root):                                                                                                                                                                                       
            return True

        if force:
            os.unlink(self.root)
            return True

        # Ensure we don't swap symlink -> exchange if we have symlink as an acceptable method.
        # This is to avoid problems switching between OSs.
        return "symlink" not in self.update_method
```

Since the last return is unreachable.  But I don't find @becker33's particularly unreadable.  I do tend to prefer eliminating possibilities without a lot of nesting.

Reply:
I would still add comments what's actually being tested, otherwise it may get turned into `lexist` in a future refactor.

Reply:
(And I still think a flat structure is easier to follow 😬)